pkg/services/ngalert/state/cache.go

```diff
@@ -137,72 +140,77 @@ func (c *cache) expandRuleLabelsAndAnnotations(ctx context.Context, alertRule *n
 	return expand(alertRule.Labels), expand(alertRule.Annotations)
 }
 
-func (c *cache) set(entry *State) {
+func (c *cache) set(newStates map[int64]map[string]*ruleStates) {
 	c.mtxStates.Lock()
 	defer c.mtxStates.Unlock()
-	if _, ok := c.states[entry.OrgID]; !ok {
-		c.states[entry.OrgID] = make(map[string]map[string]*State)
-	}
-	if _, ok := c.states[entry.OrgID][entry.AlertRuleUID]; !ok {
-		c.states[entry.OrgID][entry.AlertRuleUID] = make(map[string]*State)
-	}
-	c.states[entry.OrgID][entry.AlertRuleUID][entry.CacheId] = entry
+	c.states = newStates
```

Comment:
This is meant to swap out the full set of states, not be a partial update?

I think the function name should change to indicate that. "Set" usually means "set these keys to these values" - I wouldn't expect it to clear existing keys in the cache.
 
Maybe "clearAndSet" or "setStateTo"?

Reply:
I actually ended up renaming it :)